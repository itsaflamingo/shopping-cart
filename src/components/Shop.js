import React, {useState, useEffect} from 'react';
import '../styles/Shop.css';
import { useNavigate } from 'react-router-dom';
import Items from './Items';
import Header from './Header';

function Shop() {
  const [items, setItems] = useState([]);
  const [total, setTotal] = useState(0);
  const [cart, setCart] = useState([]);

  const addToCart = (item) => {
    let isDouble = false;
    cart.forEach((thisItem) => {
      if(thisItem.id === item.id) {
        isDouble = true;
        item.quantity = item.quantity + 1
        return;
      }
    })
    if(isDouble === false) {
      setCart(cart.concat(item))
      setTotal(((total + parseFloat(item.price))* 100)/100);
    }
  }; 

  useEffect(() => {
    loadShop();
  }, [])

  const loadShop = async () => {
    setItems(filterProducts(await shopItems()));
  } 

  const shopItems = async () => {
    const response = await fetch(`https://fakestoreapi.com/products`)
    const data = await response.json()
    data.map((item) =>  item.quantity = 1)
    return data;
  }

  const filterProducts = (products) => {
    return products.filter(product => 
      product.category === `women's clothing` || 
      product.category === `men's clothing`);
  }

  const newTotal = (price, isAdd) => {
    if(total === undefined) return setTotal(parseInt(0));
    if(isAdd === true) {
      return setTotal(((total + price)* 100)/100);
    };
    return setTotal(parseFloat((total - price).toFixed(2)));
  };

  const nav = useNavigate();

  const navHome = () => {
    nav('/')
  }

  return (
    <div id="shop">
      <Header navHome={navHome} cart={cart} newTotal={newTotal} total={total} />
      <Items items={items} addToCart={addToCart} /> 
    </div>
  );
}

export default Shop;
