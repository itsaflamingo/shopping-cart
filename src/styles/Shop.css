@import url('https://fonts.googleapis.com/css2?family=Montserrat&display=swap');
/* TOP COMPONENTS */
body {
    margin: 0;
    font-family: 'Montserrat', sans-serif;
}
.App {
    width: 100vw;
    height: 90vh;
}
#shop {
    height: auto;
    width: 100vw;
    display: flex;
    flex-direction: column;
    align-items: center;
}
/* HOMEPAGE */
#home-header {
    display: flex;
    justify-content: center;
    color: rgb(220, 220, 222);
}
#background {
    width: 100%;
    height: 100%;
    position: relative;
}
#visit-shop {
    position: absolute;
    top: 43%;
    left: 44%;
    width: 12%;
    height: 7%;
    border: 2px solid rgb(108, 118, 118);
    background-color: rgba(0, 0, 0, 0.055);
    color: white;
    font-size: 1.2rem;
}
#visit-shop:hover {
    border: 2px solid rgba(146, 164, 164, 0.781);
    background-color: rgba(146, 164, 164, 0.781);
    color: black;
}
/* SHOP HEADER */
#shop-header {
    width: 80%;
    display: flex;
    justify-content: space-between;
    align-items: center;
}
#to-home, #search {
    height: 2rem;
    width: 2rem;
}
#cart-icon {
    height: 2rem;
    width: 2rem;
    margin-left: 20%;
    position: relative;
}
#cart-quantity {
    border: 1px solid red;
    width: 0.8rem;
    height: 0.8rem;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 50%;
    background-color: red;
    color: white;
    font-weight: 900;
    font-size: 0.8rem;
    position: absolute;
    left: 70%;
}
#buttons {
    display: flex;
    width: 20%;
    justify-content: flex-end;
}
/* ITEMS */
#items {
    width: 80%;
    height: auto;
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(200px, 300px));
    justify-content: center;
    align-items: center;
    position: relative;
    margin-top: 3%;
}
.item {
    display: flex;
    flex-direction: column;
    align-items: center;
    height: 25rem;
    gap: 3rem;
}
.item-info {
    display: none;
}
.item-container:hover .item-info {
    display: flex;
    width: 70%;
    justify-content: space-between;
}
.item-container:hover #add-to-cart {
    display: flex;
    justify-content: center;
    align-items: center;
}
.image {
    height: 50%;
    width: 60%;
    display: flex;
    align-items: center;
    justify-content: center;
    max-height: 300px;
    max-width: 300px;
}
.title {
    width: 70%;
    font-weight: 900;
}
/* CART */
#cart {
    width: 20%;
    border: 1px solid black;
    position: absolute;
    top: 10%;
    left: 70%;
    z-index: 10;
    background-color: white;
    display: flex;
    flex-direction: column;
    align-items: center;
}
.cart-item {
    display: flex;
    width: 100%;
    border-bottom: 1px solid grey;
    padding-bottom: 2rem;
}
.cart-item > .cart-picture {
    justify-content: center;
}
.cart-picture {
    max-height: 8rem;
    width: 6rem;
    margin: 1rem 2rem;
}
.cart-item-info {
    display: flex;
    flex-direction: column;
    margin-right: 1rem;
}
#cart-header {
    border-bottom: 1px solid grey;
}
#checkout {
    justify-self: center;
    margin-bottom: 2rem;
}
.counter {
    display: flex;
}
.item-counter {
    width: 2rem;
}
#add-to-cart {
    display: none;
    height: 2rem;
    width: 8rem;
    background-color: rgb(0, 0, 0);
    color: white;
    font-size: 1rem;
    font-weight: 500;
}
.add, .remove {
    height: 2rem;
    width: 2rem;
    background-color: grey;
    font-size: 1.7rem;
    font-weight: 500;
    color: white;
    border: 1px solid white;
}
#total {
    margin-top: 2rem;
}
#checkout {
    height: 2rem;
    width: 8rem;
    background-color: black;
    color: white;
    text-align: center;
    font-size: 1.2rem;
}
button {
    background-color: white;
    border: none;
}
