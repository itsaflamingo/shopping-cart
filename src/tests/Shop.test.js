import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";  // optional
import Shop from '../components/Shop'
import {BrowserRouter as Router} from 'react-router-dom';
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";

// shop renders
describe('Shop component', () => {
    it('renders correct heading', () => {
        render(
            <Router>
                <Shop />
            </Router>)
        expect(screen.getByRole('heading').textContent).toMatch(/Shop/i)
    })
    it('page switches to home on button click', () => {
        render(
            <Router>
                <Shop />
            </Router>)

        const button = screen.getByRole('button', {name: 'Home'});
        act(() => {
            button.dispatchEvent(new MouseEvent('click', {bubbles: true}))
        });
        expect(window.location.pathname).toEqual('/');
    })

})
// button leads to home

// Items renders
// Items are displayed   