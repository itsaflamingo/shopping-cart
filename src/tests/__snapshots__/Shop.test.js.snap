// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shop component Renders correctly 1`] = `
<div>
  <div
    id="shop"
  >
    <div
      id="shop-header"
    >
      <h2>
        Fake Shop
      </h2>
      <div
        id="buttons"
      >
        <div
          id="search-container"
        >
          <div
            id="search"
            style="background-image: url(search.png); background-size: contain; background-repeat: no-repeat;"
          />
        </div>
        <div
          data-testid="home-icon"
          id="home-icon"
          style="background-image: url(home.png); background-size: contain; background-repeat: no-repeat;"
        />
        <div
          data-testid="cart-icon"
          id="cart-icon"
          style="background-image: url(shopping-cart.png); background-size: contain; background-repeat: no-repeat;"
        >
          <div
            id="cart-quantity"
          >
            0
          </div>
        </div>
      </div>
    </div>
    <div
      id="items"
    />
  </div>
</div>
`;
